### PR TITLE
Support MoE sigmoid scoring

### PR DESCRIPTION
**Pull request by catswe**
# Description

This enables running GLM models https://github.com/vllm-project/tpu-inference/issues/1104. I referenced this commit from Kyuyeun https://github.com/vllm-project/tpu-inference/commit/652318d9a5b6bcfd731c2ed7fd578fe975ebecd4

# Tests

image

Updated tpu-inference/tests/kernels/fused_moe_v1_test.py 

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/common/fused_moe_gmm.py [hunk 1]
```diff
@@ -361,13 +373,8 @@ def _ragged_all_to_all(operand, input_offsets, send_sizes, output_offsets,
...
     jax.jit,
-    static_argnames=(
-        "topk",
-        "renormalize",
-        "mesh",
-        "use_ep",
-        "activation",
-    ),
+    static_argnames=("topk", "renormalize", "mesh", "use_ep", "activation",
+                     "scoring_func"),
```
Comment (kyuyeunk) [inline tpu_inference/layers/common/fused_moe_gmm.py]: can you use the original formatting? changing `"scoring_func"`  to `"scoring_func",` should help formatter to automatically handle it.

File: tpu_inference/kernels/fused_moe/v1/kernel.py [hunk 2]
```diff
@@ -51,6 +51,17 @@ def broadcast_minor(src, shape):
...
 
+def scoring_fn(scoring: str, x):
+    x = x.astype(jnp.float32)
+    match scoring:
+        case "softmax":
+            return jax.nn.softmax(x, axis=-1)
+        case "sigmoid":
+            return jax.nn.sigmoid(x)
+        case _:
+            raise RuntimeError(f"Unsupported scoring function: {scoring}")
```
Comment (kyuyeunk) [inline tpu_inference/kernels/fused_moe/v1/kernel.py]: I think `NotImplementedError` is better?

**Review by kyuyeunk — COMMENTED** [review]
looks good. but requires approval from @bythew3i on changes for `fused_moe/v1/kernel.py`.

File: tpu_inference/layers/common/fused_moe_gmm.py (hunk 1 above)
Comment (catswe) [reply tpu_inference/layers/common/fused_moe_gmm.py]: Done! Thank you!

File: tpu_inference/kernels/fused_moe/v1/kernel.py (hunk 2 above)
Comment (catswe) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: Done! Based on this comment, I also took the liberty to change `activation_fn`'s `RuntimeError` to `NotImplementedError`

**Comment by bvrockwell** [comment]
This is sweet @catswe @kyuyeunk - great job!

**Comment by kyuyeunk** [comment]
> looks good. but requires approval from @bythew3i on changes for `fused_moe/v1/kernel.py`.

@bythew3i gentle ping.

File: tpu_inference/kernels/fused_moe/v1/kernel.py [hunk 3]
```diff
@@ -83,6 +95,7 @@ def ref_moe(
         *,
         renormalize_topk_logits: bool = False,
         act_fn: str = "silu",
+        scoring_func: str = "softmax",
```
Comment (bythew3i) [inline tpu_inference/kernels/fused_moe/v1/kernel.py]: Please follow the `act_fn` format to name scoring function

File: tpu_inference/kernels/fused_moe/v1/kernel.py [hunk 4]
```diff
@@ -51,6 +51,18 @@ def broadcast_minor(src, shape):
                            axis=-1)[..., :shape[-1]]
 
 
+def scoring_fn(scoring: str, x):
```
Comment (bythew3i) [inline tpu_inference/kernels/fused_moe/v1/kernel.py]: Nit: might be better to call it `get_scoring_func`

Comment (catswe) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: I renamed it to `apply_scoring_fn` for better clarity. I also took the liberty to rename `activation_fn` to `apply_act_fn` as `activation_fn` is too similar to `act_fn` @bythew3i

File: tpu_inference/kernels/fused_moe/v1/kernel.py (hunk 3 above)
Comment (catswe) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: Done

File: tpu_inference/layers/vllm/fused_moe.py [hunk 5]
```diff
@@ -121,6 +120,7 @@ def fused_moe_apply(
                     mesh=mesh,
                     use_ep=layer.use_ep,
                     activation=layer.activation,
+                    scoring_func=layer.scoring_func,
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/fused_moe.py]: nit: change this to `scoring_fn` as well so we are consistent across entire code base?

File: tpu_inference/layers/common/fused_moe_gmm.py [hunk 6]
```diff
@@ -26,6 +26,18 @@
 from tpu_inference.utils import get_mesh_shape_product
 
 
+def scoring_fn(scoring: str, x: jax.Array) -> jax.Array:
```
Comment (kyuyeunk) [inline tpu_inference/layers/common/fused_moe_gmm.py]: also, this to `apply_scoring_fn`

File: tpu_inference/layers/common/fused_moe_gmm.py [hunk 7]
```diff
@@ -26,6 +26,18 @@
...
+        case "softmax":
+            return jax.nn.softmax(x, axis=-1)
+        case "sigmoid":
+            return jax.nn.sigmoid(x)
+        case _:
+            raise NotImplementedError(
+                f"FusedMoE does not support {scoring} scoring")
+
+
 def activation_fn(activation: str, x1: jax.Array, x2: jax.Array) -> jax.Array:
```
Comment (kyuyeunk) [inline tpu_inference/layers/common/fused_moe_gmm.py]: and this to `apply_act_fn`

File: tpu_inference/kernels/fused_moe/v1/kernel.py [hunk 8]
```diff
@@ -51,6 +51,18 @@ def broadcast_minor(src, shape):
                            axis=-1)[..., :shape[-1]]
 
 
+def apply_scoring_fn(scoring_fn: str, x):
+    x = x.astype(jnp.float32)
```
Comment (bythew3i) [inline tpu_inference/kernels/fused_moe/v1/kernel.py]: why it is converted to f32, would you please elaborate?

Comment (catswe) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: As far as I know, upcasting input to fp32 is standard practice for MoE routing, for both sigmoid and softmax, especially since routing is highly sensitive to precision. This operation is computationally inexpensive and is consistent with upstream, where input is also upcasted to fp32 (see https://github.com/vllm-project/vllm/blob/06b557ecd9881289afcfa43458fb07329c955921/csrc/moe/topk_softmax_kernels.cu)

File: tpu_inference/layers/vllm/fused_moe.py (hunk 5 above)
Comment (catswe) [reply tpu_inference/layers/vllm/fused_moe.py]: since you didn't mention renaming `activation` to `act_fn`, i assume you don't want that changed and only renamed `scoring_func` to `scoring_fn`

File: tpu_inference/kernels/fused_moe/v1/kernel.py (hunk 8 above)
Comment (bythew3i) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: How much is the precision drop  if you keep original dtype?

Comment (catswe) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: I can't verify how much this would affect the model output as I'm unable to find a model small enough or currently supported to run on my TPUs where the original dtype is not already in fp32. Upcasting to `fp32` when calculating the scores is default in vLLM upstream and is computationally inexpensive. Is there a specific reason you're asking? I'm happy to investigate further as necessary!

Comment (bythew3i) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: TPU v7 supports 16-bit elementwise ops. Upcast and downcast are not free on TPU. Please use original dtype.

Comment (catswe) [reply tpu_inference/kernels/fused_moe/v1/kernel.py]: Thank you for clarifying! I updated to use the original dtype (both before softmax and sigmoid). Note that the original code also upcast to fp32 when doing softmax. Let me know if this is an issue

**Review by kyuyeunk — APPROVED** [review]

**Comment by catswe** [comment]
The newly written unit test failed since bf16 input causes a `jax.nn.sigmoid` Pallas Mosaic lowering error. I already filed a fix PR here https://github.com/jax-ml/jax/pull/34674. One workaround that works properly and passes the test is this 

```python
def apply_scoring_fn(scoring_fn: str, x):
    match scoring_fn:
        case "softmax":
            return jax.nn.softmax(x, axis=-1)
        case "sigmoid":
            # TODO(catswe): switch to jax.nn.sigmoid once bf16 Mosaic lowering bug is fixed
            ones = jnp.ones_like(x)
            return ones / (ones + jnp.exp(-x))
        case _:
            raise NotImplementedError(
                f"Unsupported scoring function: {scoring_fn}")
```

Let me know what direction we want to move towards (e.g., move forward with the workaround, or wait for the bug fix to land and upgrade Jax version)

**Comment by kyuyeunk** [comment]
> The newly written unit test failed since bf16 input causes `jax.nn.sigmoid` Pallas Mosaic lowering error. I already filed a fix here [jax-ml/jax#34674](https://github.com/jax-ml/jax/pull/34674). One workaround that works properly and passes the test is this
> 
> ```python
> def apply_scoring_fn(scoring_fn: str, x):
>     match scoring_fn:
>         case "softmax":
>             return jax.nn.softmax(x, axis=-1)
>         case "sigmoid":
>             # TODO(catswe): switch to jax.nn.sigmoid once bf16 Mosaic lowering bug is fixed
>             ones = jnp.ones_like(x)
>             return ones / (ones + jnp.exp(-x))
>         case _:
>             raise NotImplementedError(
>                 f"Unsupported scoring function: {scoring_fn}")
> ```
> 
> Let me know what direction we want to move towards

thanks for identifying this issue! I'll run some tests myself as well and get back to you if i find anything.

**Comment by catswe** [comment]
I pushed up a small change to manually implement sigmoid. `jax.nn.sigmoid` Pallas Mosaic lowering uses the same 1 / (1 + exp(-x)) formula, although there's a bug with bf16 input. I added a TODO to use `jax.nn.sigmoid` once the lowering bug is fixed and Jax version is updated

Edit: Apologies for the mass reviewers ping. Messed up my force push 😅

**Comment by kyuyeunk** [comment]
> I pushed up a small change to manually implement sigmoid. `jax.nn.sigmoid` Pallas Mosaic lowering uses the same 1 / (1 + exp(-x)) formula, although there's a bug with bf16 input. I added a TODO to use `jax.nn.sigmoid` once the lowering bug is fixed and Jax version is updated
> 
> Edit: Apologies for the mass reviewers ping. Messed up my force push 😅

just ran a quick test with latest jax version and it still triggered error. so current approach seems good to me. current approach of manually implementing sigmoid lgtm. thank you for the contribution!
